In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')
plt.rcParams['figure.figsize'] = (10,5)
plt.rcParams['figure.dpi'] = 210
sns.set_theme(style = 'darkgrid')
from sklearn.ensemble import IsolationForest

In [60]:
df = pd.read_csv(r'F:\Krishna\Excel R\Data Scientist\Assignments\EDA2\adult_with_headers.csv')
df

age          workclass  fnlwgt    education  education_num  \
0       39          State-gov   77516    Bachelors             13   
1       50   Self-emp-not-inc   83311    Bachelors             13   
2       38            Private  215646      HS-grad              9   
3       53            Private  234721         11th              7   
4       28            Private  338409    Bachelors             13   
...    ...                ...     ...          ...            ...   
32556   27            Private  257302   Assoc-acdm             12   
32557   40            Private  154374      HS-grad              9   
32558   58            Private  151910      HS-grad              9   
32559   22            Private  201490      HS-grad              9   
32560   52       Self-emp-inc  287927      HS-grad              9   

            marital_status          occupation    relationship    race  \
0            Never-married        Adm-clerical   Not-in-family   White   
1       Married-civ-spouse     Exec-managerial         Husband   White   
2                 Divorced   Handlers-cleaners   Not-in-family   White   
3       Married-civ-spouse   Handlers-cleaners         Husband   Black   
4       Married-civ-spouse      Prof-specialty            Wife   Black   
...                    ...                 ...             ...     ...   
32556   Married-civ-spouse        Tech-support            Wife   White   
32557   Married-civ-spouse   Machine-op-inspct         Husband   White   
32558              Widowed        Adm-clerical       Unmarried   White   
32559        Never-married        Adm-clerical       Own-child   White   
32560   Married-civ-spouse     Exec-managerial            Wife   White   

           sex  capital_gain  capital_loss  hours_per_week  native_country  \
0         Male          2174             0              40   United-States   
1         Male             0             0              13   United-States   
2         Male             0             0              40   United-States   
3         Male             0             0              40   United-States   
4       Female             0             0              40            Cuba   
...        ...           ...           ...             ...             ...   
32556   Female             0             0              38   United-States   
32557     Male             0             0              40   United-States   
32558   Female             0             0              40   United-States   
32559     Male             0             0              20   United-States   
32560   Female         15024             0              40   United-States   

       income  
0       <=50K  
1       <=50K  
2       <=50K  
3       <=50K  
4       <=50K  
...       ...  
32556   <=50K  
32557    >50K  
32558   <=50K  
32559   <=50K  
32560    >50K  

[32561 rows x 15 columns]

In [3]:
df.describe()

age        fnlwgt  education_num  capital_gain  capital_loss  \
count  32561.000000  3.256100e+04   32561.000000  32561.000000  32561.000000   
mean      38.581647  1.897784e+05      10.080679   1077.648844     87.303830   
std       13.640433  1.055500e+05       2.572720   7385.292085    402.960219   
min       17.000000  1.228500e+04       1.000000      0.000000      0.000000   
25%       28.000000  1.178270e+05       9.000000      0.000000      0.000000   
50%       37.000000  1.783560e+05      10.000000      0.000000      0.000000   
75%       48.000000  2.370510e+05      12.000000      0.000000      0.000000   
max       90.000000  1.484705e+06      16.000000  99999.000000   4356.000000   

       hours_per_week  
count    32561.000000  
mean        40.437456  
std         12.347429  
min          1.000000  
25%         40.000000  
50%         40.000000  
75%         45.000000  
max         99.000000

In [4]:
df.isna().sum()

age               0
workclass         0
fnlwgt            0
education         0
education_num     0
marital_status    0
occupation        0
relationship      0
race              0
sex               0
capital_gain      0
capital_loss      0
hours_per_week    0
native_country    0
income            0
dtype: int64

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       32561 non-null  object
 2   fnlwgt          32561 non-null  int64 
 3   education       32561 non-null  object
 4   education_num   32561 non-null  int64 
 5   marital_status  32561 non-null  object
 6   occupation      32561 non-null  object
 7   relationship    32561 non-null  object
 8   race            32561 non-null  object
 9   sex             32561 non-null  object
 10  capital_gain    32561 non-null  int64 
 11  capital_loss    32561 non-null  int64 
 12  hours_per_week  32561 non-null  int64 
 13  native_country  32561 non-null  object
 14  income          32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


There are no null values.

In [6]:
df.dtypes

age                int64
workclass         object
fnlwgt             int64
education         object
education_num      int64
marital_status    object
occupation        object
relationship      object
race              object
sex               object
capital_gain       int64
capital_loss       int64
hours_per_week     int64
native_country    object
income            object
dtype: object

In [7]:
df

age          workclass  fnlwgt    education  education_num  \
0       39          State-gov   77516    Bachelors             13   
1       50   Self-emp-not-inc   83311    Bachelors             13   
2       38            Private  215646      HS-grad              9   
3       53            Private  234721         11th              7   
4       28            Private  338409    Bachelors             13   
...    ...                ...     ...          ...            ...   
32556   27            Private  257302   Assoc-acdm             12   
32557   40            Private  154374      HS-grad              9   
32558   58            Private  151910      HS-grad              9   
32559   22            Private  201490      HS-grad              9   
32560   52       Self-emp-inc  287927      HS-grad              9   

            marital_status          occupation    relationship    race  \
0            Never-married        Adm-clerical   Not-in-family   White   
1       Married-civ-spouse     Exec-managerial         Husband   White   
2                 Divorced   Handlers-cleaners   Not-in-family   White   
3       Married-civ-spouse   Handlers-cleaners         Husband   Black   
4       Married-civ-spouse      Prof-specialty            Wife   Black   
...                    ...                 ...             ...     ...   
32556   Married-civ-spouse        Tech-support            Wife   White   
32557   Married-civ-spouse   Machine-op-inspct         Husband   White   
32558              Widowed        Adm-clerical       Unmarried   White   
32559        Never-married        Adm-clerical       Own-child   White   
32560   Married-civ-spouse     Exec-managerial            Wife   White   

           sex  capital_gain  capital_loss  hours_per_week  native_country  \
0         Male          2174             0              40   United-States   
1         Male             0             0              13   United-States   
2         Male             0             0              40   United-States   
3         Male             0             0              40   United-States   
4       Female             0             0              40            Cuba   
...        ...           ...           ...             ...             ...   
32556   Female             0             0              38   United-States   
32557     Male             0             0              40   United-States   
32558   Female             0             0              40   United-States   
32559     Male             0             0              20   United-States   
32560   Female         15024             0              40   United-States   

       income  
0       <=50K  
1       <=50K  
2       <=50K  
3       <=50K  
4       <=50K  
...       ...  
32556   <=50K  
32557    >50K  
32558   <=50K  
32559   <=50K  
32560    >50K  

[32561 rows x 15 columns]

All datatypes are correct.

we can remove some of the not much important features from the dataset and for the same we will find relation between the target column and all other independent columns. For that we will use PPS (Predictive Power Score) method.

In [10]:
!pip install ppscore

In [11]:
import ppscore as ps

In [12]:
ps.score(df, x = 'age', y = 'income') #to know the relationship between two columns 

{'x': 'age',
 'y': 'income',
 'ppscore': 0.005415334901707637,
 'case': 'classification',
 'is_valid_score': True,
 'metric': 'weighted F1',
 'baseline_score': 0.6531153390987711,
 'model_score': 0.6549938357098165,
 'model': DecisionTreeClassifier()}

We can see their is almost no relation between age and income.

In [13]:
feature_sel = ps.matrix(df)      #to know the relationship between every possible pair of a dataset

In [14]:
feature_sel

x               y   ppscore            case  is_valid_score  \
0       age             age  1.000000  predict_itself            True   
1       age       workclass  0.011232  classification            True   
2       age          fnlwgt  0.000000      regression            True   
3       age       education  0.052315  classification            True   
4       age   education_num  0.000000      regression            True   
..      ...             ...       ...             ...             ...   
220  income    capital_gain  0.000000      regression            True   
221  income    capital_loss  0.000000      regression            True   
222  income  hours_per_week  0.000000      regression            True   
223  income  native_country  0.000000  classification            True   
224  income          income  1.000000  predict_itself            True   

                  metric  baseline_score   model_score  \
0                   None        0.000000      1.000000   
1            weighted F1        0.579088      0.583816   
2    mean absolute error    75872.186200  77535.141544   
3            weighted F1        0.201200      0.242989   
4    mean absolute error        1.853000      1.898306   
..                   ...             ...           ...   
220  mean absolute error     1093.884000   1760.682115   
221  mean absolute error       94.942600    176.261353   
222  mean absolute error        7.656400      8.097596   
223          weighted F1        0.841082      0.841082   
224                 None        0.000000      1.000000   

                        model  
0                        None  
1    DecisionTreeClassifier()  
2     DecisionTreeRegressor()  
3    DecisionTreeClassifier()  
4     DecisionTreeRegressor()  
..                        ...  
220   DecisionTreeRegressor()  
221   DecisionTreeRegressor()  
222   DecisionTreeRegressor()  
223  DecisionTreeClassifier()  
224                      None  

[225 rows x 9 columns]

In [15]:
feature_sel[feature_sel['y'] == 'income'] 

x       y   ppscore            case  is_valid_score  \
14              age  income  0.005415  classification            True   
29        workclass  income  0.094056  classification            True   
44           fnlwgt  income  0.000000  classification            True   
59        education  income  0.243135  classification            True   
74    education_num  income  0.243135  classification            True   
89   marital_status  income  0.000000  classification            True   
104      occupation  income  0.092410  classification            True   
119    relationship  income  0.000000  classification            True   
134            race  income  0.000000  classification            True   
149             sex  income  0.000000  classification            True   
164    capital_gain  income  0.297123  classification            True   
179    capital_loss  income  0.141755  classification            True   
194  hours_per_week  income  0.047278  classification            True   
209  native_country  income  0.009409  classification            True   
224          income  income  1.000000  predict_itself            True   

          metric  baseline_score  model_score                     model  
14   weighted F1        0.653115     0.654994  DecisionTreeClassifier()  
29   weighted F1        0.653115     0.685742  DecisionTreeClassifier()  
44   weighted F1        0.653115     0.644900  DecisionTreeClassifier()  
59   weighted F1        0.653115     0.737455  DecisionTreeClassifier()  
74   weighted F1        0.653115     0.737455  DecisionTreeClassifier()  
89   weighted F1        0.653115     0.653017  DecisionTreeClassifier()  
104  weighted F1        0.653115     0.685171  DecisionTreeClassifier()  
119  weighted F1        0.653115     0.653115  DecisionTreeClassifier()  
134  weighted F1        0.653115     0.653115  DecisionTreeClassifier()  
149  weighted F1        0.653115     0.653115  DecisionTreeClassifier()  
164  weighted F1        0.653115     0.756183  DecisionTreeClassifier()  
179  weighted F1        0.653115     0.702288  DecisionTreeClassifier()  
194  weighted F1        0.653115     0.669515  DecisionTreeClassifier()  
209  weighted F1        0.653115     0.656379  DecisionTreeClassifier()  
224         None        0.000000     1.000000                      None

As per above, we can see that for predicting if income is >50K or not, we definitely don't need columns like age, fnlwgt, marital_status, relationship, race, sex etc... So it would be better to remove those features from our dataset to get higher accuracy results.

In [61]:
df.drop(columns = ['age', 'fnlwgt', 'marital_status', 'relationship', 'race', 'sex'], inplace = True)

In [62]:
df

workclass    education  education_num          occupation  \
0              State-gov    Bachelors             13        Adm-clerical   
1       Self-emp-not-inc    Bachelors             13     Exec-managerial   
2                Private      HS-grad              9   Handlers-cleaners   
3                Private         11th              7   Handlers-cleaners   
4                Private    Bachelors             13      Prof-specialty   
...                  ...          ...            ...                 ...   
32556            Private   Assoc-acdm             12        Tech-support   
32557            Private      HS-grad              9   Machine-op-inspct   
32558            Private      HS-grad              9        Adm-clerical   
32559            Private      HS-grad              9        Adm-clerical   
32560       Self-emp-inc      HS-grad              9     Exec-managerial   

       capital_gain  capital_loss  hours_per_week  native_country  income  
0              2174             0              40   United-States   <=50K  
1                 0             0              13   United-States   <=50K  
2                 0             0              40   United-States   <=50K  
3                 0             0              40   United-States   <=50K  
4                 0             0              40            Cuba   <=50K  
...             ...           ...             ...             ...     ...  
32556             0             0              38   United-States   <=50K  
32557             0             0              40   United-States    >50K  
32558             0             0              40   United-States   <=50K  
32559             0             0              20   United-States   <=50K  
32560         15024             0              40   United-States    >50K  

[32561 rows x 9 columns]

Encoding

In [24]:
df['workclass'].unique()

array([' State-gov', ' Self-emp-not-inc', ' Private', ' Federal-gov',
       ' Local-gov', ' ?', ' Self-emp-inc', ' Without-pay',
       ' Never-worked'], dtype=object)

In [25]:
df['workclass'].nunique()

9

In [26]:
df['education'].nunique()

16

In [28]:
df['occupation'].nunique()

15

In [32]:
df['native_country'].nunique()

42

In [33]:
df['income'].nunique()

2

We can see we have so many columns with Text values , so first of all ,  we will encode them. For independent columns having less than 5 unique values, we will use one hot encoding. For independent columns having greater than 5 unique values, we will use Ordinal Encoding and for target column we will use label Encoding.

*  OHE (One Hot Encoding) : Pros : easy to use , used for independent columns
                         Cons : For columns having more unique values, it creates so many new columns
*  Label Encoding   : Pros : doesnt matter how many unique values are there, it encode them in a single column
                   Cons : used only for target columns 
*  Ordinal Encoding    : Pros : easy to use, doesnt create so many columns, encode them in a single column only
                      Cons  : used only for independent columns 

In [63]:
from sklearn.preprocessing import OrdinalEncoder

In [64]:
df.select_dtypes(object)

workclass    education          occupation  native_country  \
0              State-gov    Bachelors        Adm-clerical   United-States   
1       Self-emp-not-inc    Bachelors     Exec-managerial   United-States   
2                Private      HS-grad   Handlers-cleaners   United-States   
3                Private         11th   Handlers-cleaners   United-States   
4                Private    Bachelors      Prof-specialty            Cuba   
...                  ...          ...                 ...             ...   
32556            Private   Assoc-acdm        Tech-support   United-States   
32557            Private      HS-grad   Machine-op-inspct   United-States   
32558            Private      HS-grad        Adm-clerical   United-States   
32559            Private      HS-grad        Adm-clerical   United-States   
32560       Self-emp-inc      HS-grad     Exec-managerial   United-States   

       income  
0       <=50K  
1       <=50K  
2       <=50K  
3       <=50K  
4       <=50K  
...       ...  
32556   <=50K  
32557    >50K  
32558   <=50K  
32559   <=50K  
32560    >50K  

[32561 rows x 5 columns]

In [65]:
cat_cols = df.drop(columns = 'income').select_dtypes(object).columns 
cat_cols

Index(['workclass', 'education', 'occupation', 'native_country'], dtype='object')

In [66]:
df[cat_cols]

workclass    education          occupation  native_country
0              State-gov    Bachelors        Adm-clerical   United-States
1       Self-emp-not-inc    Bachelors     Exec-managerial   United-States
2                Private      HS-grad   Handlers-cleaners   United-States
3                Private         11th   Handlers-cleaners   United-States
4                Private    Bachelors      Prof-specialty            Cuba
...                  ...          ...                 ...             ...
32556            Private   Assoc-acdm        Tech-support   United-States
32557            Private      HS-grad   Machine-op-inspct   United-States
32558            Private      HS-grad        Adm-clerical   United-States
32559            Private      HS-grad        Adm-clerical   United-States
32560       Self-emp-inc      HS-grad     Exec-managerial   United-States

[32561 rows x 4 columns]

In [67]:
oe = OrdinalEncoder()           

df[cat_cols] = oe.fit_transform(df[cat_cols])

In [68]:
df[cat_cols]

workclass  education  occupation  native_country
0            7.0        9.0         1.0            39.0
1            6.0        9.0         4.0            39.0
2            4.0       11.0         6.0            39.0
3            4.0        1.0         6.0            39.0
4            4.0        9.0        10.0             5.0
...          ...        ...         ...             ...
32556        4.0        7.0        13.0            39.0
32557        4.0       11.0         7.0            39.0
32558        4.0       11.0         1.0            39.0
32559        4.0       11.0         1.0            39.0
32560        5.0       11.0         4.0            39.0

[32561 rows x 4 columns]

In [69]:
df

workclass  education  education_num  occupation  capital_gain  \
0            7.0        9.0             13         1.0          2174   
1            6.0        9.0             13         4.0             0   
2            4.0       11.0              9         6.0             0   
3            4.0        1.0              7         6.0             0   
4            4.0        9.0             13        10.0             0   
...          ...        ...            ...         ...           ...   
32556        4.0        7.0             12        13.0             0   
32557        4.0       11.0              9         7.0             0   
32558        4.0       11.0              9         1.0             0   
32559        4.0       11.0              9         1.0             0   
32560        5.0       11.0              9         4.0         15024   

       capital_loss  hours_per_week  native_country  income  
0                 0              40            39.0   <=50K  
1                 0              13            39.0   <=50K  
2                 0              40            39.0   <=50K  
3                 0              40            39.0   <=50K  
4                 0              40             5.0   <=50K  
...             ...             ...             ...     ...  
32556             0              38            39.0   <=50K  
32557             0              40            39.0    >50K  
32558             0              40            39.0   <=50K  
32559             0              20            39.0   <=50K  
32560             0              40            39.0    >50K  

[32561 rows x 9 columns]

We have encoded all the independent columns with One hot encoding and Ordinal Encoder. Now, let's encode Target Column with Label Encoder.

In [70]:
from sklearn.preprocessing import LabelEncoder

In [71]:
le = LabelEncoder()

In [72]:
df['income'] = le.fit_transform(df['income'])

In [73]:
df

workclass  education  education_num  occupation  capital_gain  \
0            7.0        9.0             13         1.0          2174   
1            6.0        9.0             13         4.0             0   
2            4.0       11.0              9         6.0             0   
3            4.0        1.0              7         6.0             0   
4            4.0        9.0             13        10.0             0   
...          ...        ...            ...         ...           ...   
32556        4.0        7.0             12        13.0             0   
32557        4.0       11.0              9         7.0             0   
32558        4.0       11.0              9         1.0             0   
32559        4.0       11.0              9         1.0             0   
32560        5.0       11.0              9         4.0         15024   

       capital_loss  hours_per_week  native_country  income  
0                 0              40            39.0       0  
1                 0              13            39.0       0  
2                 0              40            39.0       0  
3                 0              40            39.0       0  
4                 0              40             5.0       0  
...             ...             ...             ...     ...  
32556             0              38            39.0       0  
32557             0              40            39.0       1  
32558             0              40            39.0       0  
32559             0              20            39.0       0  
32560             0              40            39.0       1  

[32561 rows x 9 columns]

Now, let's detect Outliers and remove them to get better accuracy. We will use Random Forest method for the same.

In [74]:
iso = IsolationForest()
iso.fit(df)              
ypred = iso.predict(df)    
ypred 

array([ 1,  1,  1, ...,  1,  1, -1])

In [75]:
len(ypred)

32561

In [76]:
df['Outliers'] = ypred

df

workclass  education  education_num  occupation  capital_gain  \
0            7.0        9.0             13         1.0          2174   
1            6.0        9.0             13         4.0             0   
2            4.0       11.0              9         6.0             0   
3            4.0        1.0              7         6.0             0   
4            4.0        9.0             13        10.0             0   
...          ...        ...            ...         ...           ...   
32556        4.0        7.0             12        13.0             0   
32557        4.0       11.0              9         7.0             0   
32558        4.0       11.0              9         1.0             0   
32559        4.0       11.0              9         1.0             0   
32560        5.0       11.0              9         4.0         15024   

       capital_loss  hours_per_week  native_country  income  Outliers  
0                 0              40            39.0       0         1  
1                 0              13            39.0       0         1  
2                 0              40            39.0       0         1  
3                 0              40            39.0       0         1  
4                 0              40             5.0       0         1  
...             ...             ...             ...     ...       ...  
32556             0              38            39.0       0         1  
32557             0              40            39.0       1         1  
32558             0              40            39.0       0         1  
32559             0              20            39.0       0         1  
32560             0              40            39.0       1        -1  

[32561 rows x 10 columns]

In [77]:
data = df[df['Outliers'] == 1]
data              #creating a new dataset without Outliers

workclass  education  education_num  occupation  capital_gain  \
0            7.0        9.0             13         1.0          2174   
1            6.0        9.0             13         4.0             0   
2            4.0       11.0              9         6.0             0   
3            4.0        1.0              7         6.0             0   
4            4.0        9.0             13        10.0             0   
...          ...        ...            ...         ...           ...   
32555        4.0       15.0             10        11.0             0   
32556        4.0        7.0             12        13.0             0   
32557        4.0       11.0              9         7.0             0   
32558        4.0       11.0              9         1.0             0   
32559        4.0       11.0              9         1.0             0   

       capital_loss  hours_per_week  native_country  income  Outliers  
0                 0              40            39.0       0         1  
1                 0              13            39.0       0         1  
2                 0              40            39.0       0         1  
3                 0              40            39.0       0         1  
4                 0              40             5.0       0         1  
...             ...             ...             ...     ...       ...  
32555             0              40            39.0       0         1  
32556             0              38            39.0       0         1  
32557             0              40            39.0       1         1  
32558             0              40            39.0       0         1  
32559             0              20            39.0       0         1  

[27235 rows x 10 columns]

We can see , we have removed all the outliers from our dataset and now from 32561, we have only 27235 no. of rows remaining, which means there were almost 5326 outliers.

And now, We have removed unnessecery columns and outliers from our dataset. Now, we can scale our data to make it ready to build a model.

In [82]:
x = data.drop(columns = ['income'])
print(x)
y = data['income']
y

       workclass  education  education_num  occupation  capital_gain  \
0            7.0        9.0             13         1.0          2174   
1            6.0        9.0             13         4.0             0   
2            4.0       11.0              9         6.0             0   
3            4.0        1.0              7         6.0             0   
4            4.0        9.0             13        10.0             0   
...          ...        ...            ...         ...           ...   
32555        4.0       15.0             10        11.0             0   
32556        4.0        7.0             12        13.0             0   
32557        4.0       11.0              9         7.0             0   
32558        4.0       11.0              9         1.0             0   
32559        4.0       11.0              9         1.0             0   

       capital_loss  hours_per_week  native_country  Outliers  
0                 0              40            39.0         1  
1      

0        0
1        0
2        0
3        0
4        0
        ..
32555    0
32556    0
32557    1
32558    0
32559    0
Name: income, Length: 27235, dtype: int32

In [83]:
from sklearn.model_selection import train_test_split

xtrain, xtest, ytrain, ytest = train_test_split(x,y, test_size = 0.2, random_state = 42)


In [84]:
xtrain

workclass  education  education_num  occupation  capital_gain  \
21793        4.0       11.0              9         7.0             0   
3864         7.0       11.0              9         3.0             0   
27002        4.0       11.0              9         7.0             0   
1207         4.0        9.0             13        12.0             0   
4251         4.0       11.0              9        13.0             0   
...          ...        ...            ...         ...           ...   
25766        2.0        9.0             13        10.0             0   
6467         4.0        0.0              6        14.0             0   
1006         4.0       11.0              9         7.0             0   
18863        7.0       11.0              9         3.0             0   
28244        4.0        9.0             13         4.0          8614   

       capital_loss  hours_per_week  native_country  Outliers  
21793          1741              42            39.0         1  
3864              0              40            39.0         1  
27002             0              40            40.0         1  
1207              0              42            39.0         1  
4251              0              40            39.0         1  
...             ...             ...             ...       ...  
25766             0              40            30.0         1  
6467              0              20            39.0         1  
1006              0              40            39.0         1  
18863             0              40            39.0         1  
28244             0              40            39.0         1  

[21788 rows x 9 columns]

In [85]:
ytrain

21793    0
3864     1
27002    1
1207     1
4251     0
        ..
25766    1
6467     0
1006     0
18863    0
28244    1
Name: income, Length: 21788, dtype: int32

In [86]:
xtest

workclass  education  education_num  occupation  capital_gain  \
24528        6.0       11.0              9         3.0          4064   
8429         0.0       11.0              9         0.0             0   
2648         4.0       15.0             10         1.0          4101   
14002        6.0       11.0              9         8.0             0   
24086        4.0       11.0              9         7.0             0   
...          ...        ...            ...         ...           ...   
21848        4.0       15.0             10         5.0             0   
14010        4.0        9.0             13         1.0          4687   
17745        4.0        8.0             11        10.0          7298   
12803        4.0       15.0             10         4.0             0   
1763         4.0        0.0              6         7.0             0   

       capital_loss  hours_per_week  native_country  Outliers  
24528             0              40            39.0         1  
8429              0              25            39.0         1  
2648              0              50            39.0         1  
14002             0              28            39.0         1  
24086             0              40            39.0         1  
...             ...             ...             ...       ...  
21848             0              15            39.0         1  
14010             0              35            39.0         1  
17745             0              50            39.0         1  
12803             0              40            39.0         1  
1763              0              45            39.0         1  

[5447 rows x 9 columns]

In [87]:
ytest

24528    0
8429     0
2648     0
14002    0
24086    0
        ..
21848    0
14010    1
17745    1
12803    0
1763     0
Name: income, Length: 5447, dtype: int32

We have scaled the data and finally our data is ready to build the model.